clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java

```diff
@@ -0,0 +1,348 @@
+     * Therefore, we must search forward in the array until we hit a null, before
+     * concluding that the element is not present.
+     *
+     * @param example   The element to match.
+     * @return          The match index, or INVALID_INDEX if no match was found.
+     */
+    private int findIndex(E example) {
+        int slot = slot(elements, example);
+        for (int seen = 0; seen < elements.length; seen++) {
+            Element element = elements[slot];
+            if (element == null) {
+                return INVALID_INDEX;
+            }
+            if (element.equals(example)) {
+                return slot;
+            }
+            slot = (slot + 1) % elements.length;
+        }
+        return INVALID_INDEX;
+    }
+
+    /**
+     * Find the element which equals() the given example element.
+     *
+     * @param example   The example element.
+     * @return          Null if no element was found; the element, otherwise.
+     */
+    public E find(E example) {
+        int index = findIndex(example);
+        if (index == INVALID_INDEX) {
+            return null;
+        }
+        return (E) elements[index];
+    }
+
+    /**
+     * Returns the number of elements in the set.
+     */
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        E example = null;
+        try {
+            example = (E) o;
+        } catch (ClassCastException e) {
+            return false;
+        }
+        return find(example) != null;
+    }
+
+    @Override
+    public boolean add(E newElement) {
+        if ((size + 1) >= elements.length / 2) {
+            // Avoid using even-sized capacities, to get better key distribution.
+            changeCapacity((2 * elements.length) + 1);
+        }
+        int slot = addInternal(newElement, elements);
+        if (slot >= 0) {
+            addToListTail(head, elements, slot);
+            size++;
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Adds a new element to the appropriate place in the elements array.
+     *
+     * @param newElement    The new element to add.
+     * @param addElements   The elements array.
+     * @return              The index at which the element was inserted, or INVALID_INDEX
+     *                      if the element could not be inserted because there was already
+     *                      an equivalent element.
+     */
+    private static int addInternal(Element newElement, Element[] addElements) {
+        int slot = slot(addElements, newElement);
+        for (int seen = 0; seen < addElements.length; seen++) {
+            Element element = addElements[slot];
+            if (element == null) {
+                addElements[slot] = newElement;
+                return slot;
+            }
+            if (element.equals(newElement)) {
+                return INVALID_INDEX;
```

Comment:
Hmm.  It's probably good to keep the existing add function as-is, beacuse then we implement the familiar Java contract for `java.util.Set`.  However, let me add a new method called `mustAdd` which throws an exception if the addition could not be performed, so that we catch any bugs here.